godel_surface_detection/src/segmentation/surface_segmentation.cpp

```diff
@@ -453,70 +412,49 @@ void SurfaceSegmentation::getBoundaryTrajectory(std::vector<pcl::IndicesPtr> &bo
-  chull.setAlpha(500.0);
-  chull.reconstruct(mesh);
-
-  return mesh.polygons.size() > 0;
-}
-
-
-void SurfaceSegmentation::getBoundBoundaryHalfEdges (const Mesh &mesh,
-                                                     std::vector <Mesh::HalfEdgeIndices>& boundary_he_collection,
-                                                     const size_t expected_size)
-{
-
-  boundary_he_collection.clear ();
-
-  HalfEdgeIndices boundary_he; boundary_he.reserve (expected_size);
-  std::vector <bool> visited (mesh.sizeEdges (), false);
-  IHEAFC circ, circ_end;
-
-  for (HalfEdgeIndex i (0); i<HalfEdgeIndex (mesh.sizeHalfEdges ()); ++i)
-  {
-    if (mesh.isBoundary (i) && !visited [pcl::geometry::toEdgeIndex (i).get ()])
-    {
-      boundary_he.clear ();
-
-      circ     = mesh.getInnerHalfEdgeAroundFaceCirculator (i);
-      circ_end = circ;
-      do
-      {
-      visited [pcl::geometry::toEdgeIndex (circ.getTargetIndex ()).get ()] = true;
-      boundary_he.push_back (circ.getTargetIndex ());
-      } while ( ++circ != circ_end);
-
-      boundary_he_collection.push_back (boundary_he);
-    }
-  }
-}
-
-
-/** @brief remove any NAN points, otherwise many algorityms fail */
+/** @brief remove any NAN points, otherwise many algorithms fail */
 void SurfaceSegmentation::removeNans()
 {
-  pcl::PointCloud<pcl::PointXYZRGB>::Ptr nonans_cloud(new pcl::PointCloud<pcl::PointXYZRGB>);
-  nonans_cloud->is_dense = false;
   std::vector<int> indices;
-  pcl::removeNaNFromPointCloud (*input_cloud_, *nonans_cloud, indices);
-  setInputCloud(nonans_cloud);
+  pcl::removeNaNFromPointCloud (*input_cloud_, *input_cloud_, indices);
 }
 
 
 /** @brief compute the normals and store in normals_, this is requried for both segmentation and meshing*/
 void SurfaceSegmentation::computeNormals()
 {
-  normals_->points.clear();
-  // Estimate the normals
-  pcl::NormalEstimation<pcl::PointXYZRGB, pcl::Normal> ne;
+  // Determine the number of available cores
+  pcl::NormalEstimationOMP<pcl::PointXYZRGB, pcl::Normal> ne;
+  int nr_cores = std::thread::hardware_concurrency();
+  ne.setNumberOfThreads(nr_cores);
+
+  // Configure parameters
   ne.setInputCloud (input_cloud_);
   pcl::search::KdTree<pcl::PointXYZRGB>::Ptr tree (new pcl::search::KdTree<pcl::PointXYZRGB> ());
   ne.setSearchMethod (tree);
-  //  ne.setRadiusSearch (radius_);
   ne.setKSearch (100);
+
+  // Estimate the normals
   ne.compute (*normals_);
 }
+
+
+void SurfaceSegmentation::getSurfaceClouds(std::vector<pcl::PointCloud<pcl::PointXYZ>::Ptr> &surface_clouds)
+{
+  surface_clouds.clear();
+  pcl::PointCloud<pcl::PointXYZ>::Ptr segment_cloud_ptr =
+      pcl::PointCloud<pcl::PointXYZ>::Ptr(new pcl::PointCloud<pcl::PointXYZ>());
+
+  for (const auto& cluster : clusters_)
+  {
+    pcl::PointIndices indices = cluster;
+    if (indices.indices.size() == 0)
+      continue;
+
+    if (indices.indices.size() >= MIN_CLUSTER_SIZE)
+    {
+      pcl::copyPointCloud(*input_cloud_, indices, *segment_cloud_ptr);
```

Comment:
Does this reset `segment_cloud_ptr`? Otherwise it looks like we're just pushing a bunch of the same pointer into the vector and all of them will look at the last item.